lib/src/charts/widgets/bar_chart.dart

```diff
@@ -59,81 +55,51 @@ class BarChart extends StatefulWidget {
-      } else {
-        for (final data in yValue) {
-          minV = min(minV, data.yValue.toDouble());
-          maxV = max(maxV, data.yValue.toDouble());
-        }
-      }
-
-      calculatedMinYValue = min(calculatedMinYValue, minV);
-      calculatedMaxYValue = max(calculatedMaxYValue, maxV);
-      _maxBarsInGroup = max(_maxBarsInGroup, yValue.length);
-    }
-
-    // Finally we will collect our user provided min/max values
-    // and the ones that we have calculated
-    var maxYValue = widget.maxYValue ?? calculatedMaxYValue;
-    var minYValue = calculatedMinYValue;
-
-    var unit = (widget.chartStyle ?? defaultCartesianChartStyle)
-        .gridStyle!
-        .yUnitValue!;
-    var maxYRange = maxYValue;
-    while (maxYRange % unit != 0) {
-      maxYRange++;
-    }
-
-    // We need to check for negative y values
-    var minYRange = minYValue;
-    if (minYValue.isNegative) {
-      while (minYRange % unit != 0) {
-        minYRange--;
-      }
-    } else {
-      // No negative y values, so min will be zero
-      minYRange = 0.0;
-    }
-
     // Now we can provide the chart details to the observer
     _controller = CartesianController(
-      data: [widget.data],
-      animateOnLoad: widget.animateOnLoad,
-      autoAnimate: widget.autoAnimate,
-      animation: AnimationController(
-        duration: const Duration(seconds: 1),
-        vsync: this,
-      ),
-      minValue: minYValue,
-      maxValue: maxYValue,
-      maxXRange: calculatedMaxXValue,
-      minXRange: calculatedMinXValue,
-      maxYRange: maxYRange,
-      minYRange: minYRange,
-    );
+        data: [widget.data],
+        animateOnLoad: widget.animateOnLoad,
+        autoAnimate: widget.autoAnimate,
+        animation: AnimationController(
+          duration: const Duration(seconds: 1),
+          vsync: this,
+        ),
+        rangeConstraints: (c) {
+          // For Bar charts, we normally don't consider x values, be it +ve or -ve
+          c.maxXValue = widget.data.barData.length.toDouble();
+          // Finally we will collect our user provided min/max values
+          // and the ones that we have calculated
+          c.maxXRange = c.maxXValue;
+          c.minXRange = c.minXValue;
+          var minYValue = c.minYValue;
+
+          var yUnit = (widget.chartStyle ?? defaultCartesianChartStyle)
+              .gridStyle!
+              .yUnitValue!;
+
+          var maxYRange = widget.maxYValue ?? c.maxYValue;
+          while (maxYRange % yUnit != 0) {
+            maxYRange++;
+          }
+          c.maxYRange = maxYRange;
+
+          // We need to check for negative y values
+          var minYRange = minYValue;
+          if (minYValue.isNegative) {
+            while (minYRange % yUnit != 0) {
+              minYRange--;
+            }
+          } else {
+            // No negative y values, so min will be zero
+            minYRange = 0.0;
+          }
+          c.minYRange = minYRange;
+        });
```

Comment:
This callback could be helpful to account for chart specific bias in calculation of final constraints.
But there are a few problems here
1. It's not clear that this callback needs to update the `range` values. It's also not clear what's the difference between `maxXValue` vs `maxXRange`.
2. Allowing mutation of the controller from outside will lead to difficult to trace bugs. Any mutation of variables that belong to the controller should happen within the controller.

Here's what I propose:
1. Add clear documentation of what is a `value` and what is a `range`.
2. Create a data class that will contain the data needed to calculate the `range` variables (min max `value` variables). Let's call that `RangeCalculationContext`
3. Create another data class that will represent the final calculated `range` variables. Let's call that `RangeCalculationResult`
4. Rename `rangeConstraints` to `calculateRange`. We will follow the range naming convention for now, but if later we need this to do more than range calc, we can update the naming and data classes accordingly.
5. Update the type of `calculateRange` to be `RangeCalculationResult Function(RangeCalculationContext context)`
6. In the controller, you will call `calculateRange` giving it the `RangeCalculationContext`. `calculateRange` will return you `RangeCalculationResult`. Using the result you will update the range variables inside the controller itself.